src/passes/StackIR.cpp

```diff
@@ -0,0 +1,391 @@
+        values.push_back(i);
+      }
+    }
+  }
+
+  // There may be unnecessary blocks we can remove: blocks
+  // without branches to them are always ok to remove.
+  // TODO: a branch to a block in an if body can become
+  //       a branch to that if body
+  void removeUnneededBlocks() {
+    for (auto*& inst : insts) {
+      if (!inst) continue;
+      if (auto* block = inst->origin->dynCast<Block>()) {
+        if (!BranchUtils::BranchSeeker::hasNamed(block, block->name)) {
+          // TODO optimize, maybe run remove-unused-names
+          inst = nullptr;
+        }
+      }
+    }
+  }
+
+  // Utilities.
+
+  // A control flow "barrier" - a point where stack machine
+  // unreachability ends.
+  bool isControlFlowBarrier(StackInst* inst) {
+    switch (inst->op) {
+      case StackInst::BlockEnd:
+      case StackInst::IfElse:
+      case StackInst::IfEnd:
+      case StackInst::LoopEnd: {
+        return true;
+      }
+      default: {
+        return false;
+      }
+    }
+  }
+
+  // A control flow beginning.
+  bool isControlFlowBegin(StackInst* inst) {
+    switch (inst->op) {
+      case StackInst::BlockBegin:
+      case StackInst::IfBegin:
+      case StackInst::LoopBegin: {
+        return true;
+      }
+      default: {
+        return false;
+      }
+    }
+  }
+
+  // A control flow ending.
+  bool isControlFlowEnd(StackInst* inst) {
+    switch (inst->op) {
+      case StackInst::BlockEnd:
+      case StackInst::IfEnd:
+      case StackInst::LoopEnd: {
+        return true;
+      }
+      default: {
+        return false;
+      }
+    }
+  }
+
+  bool isControlFlow(StackInst* inst) {
+    return inst->op != StackInst::Basic;
+  }
+
+  // Remove the instruction at index i. If the instruction
+  // is control flow, and so has been expanded to multiple
+  // instructions, remove them as well.
+  void removeAt(Index i) {
+    auto* inst = insts[i];
+    insts[i] = nullptr;
+    if (inst->op == StackInst::Basic) {
+      return; // that was it
+    }
+    auto* origin = inst->origin;
+    while (1) {
+      i++;
+      assert(i < insts.size());
+      inst = insts[i];
+      insts[i] = nullptr;
+      if (inst && inst->origin == origin && isControlFlowEnd(inst)) {
+        return; // that's it, we removed it all
+      }
+    }
+  }
+
+  Index getNumConsumedValues(StackInst* inst) {
+    if (isControlFlow(inst)) {
+      // If consumes 1; that's it.
+      if (inst->op == StackInst::IfBegin) {
+        return 1;
+      }
+      return 0;
+    }
+    // Otherwise, for basic instructions, just count the expression children.
+    return ChildIterator(inst->origin).children.size();
+  }
+};
+
+struct OptimizeStackIR : public WalkerPass<PostWalker<OptimizeStackIR>> {
+  bool isFunctionParallel() override { return true; }
+
+  Pass* create() override { return new OptimizeStackIR; }
+
+  bool modifiesBinaryenIR() override { return false; }
+
+  void doWalkFunction(Function* func) {
+    if (!func->stackIR) {
```

Comment:
Should this check live in StackIROptimizer itself? Or should `StackIROptimizer(...).run()` never be a no-op, at which point we should maybe add an assert to `StackIROptimizer` itself.

Reply:
It could be in both places I guess. I like the idea that `StackIROptimizer` operates on valid input, containing stack IR, and the pass is valid to run if there is or isn't. Yeah, makes sense to add an assert inside StackIROptimizer, I'll add that.